src/services/counters/CounterService.ts

```diff
@@ -0,0 +1,176 @@
+            if (traversingCounterObject) {
+              let counter: any = deepValue(response, trail);
+              let type: string = trail[0];
+              if (isNotNullOrUndefined(counter) && type === "strong-counter") {
+                counters.push(this.createStrongCounter(counter.name, counter.storage, counter["initial-value"],
+                  counter.value, counter["lower-bound"], counter["upper-bound"], counter));
+              } else if (isNotNullOrUndefined(counter) && type === "weak-counter") {
+                counters.push(this.createWeakCounter(counter.name, counter.storage, counter["initial-value"],
+                  counter.value, counter.concurrency, counter));
+              }
+            }
+          }, trail);
+          return counters;
+        }).then(counters => {
+        deferred.resolve(counters);
+      })
+    });
+    return deferred.promise;
+  }
+
+  update(counter: ICounter, profile: string, container: string): ng.IPromise<any> {
+    return this.save(counter, profile, container, ["is-new-node"]);
+  }
+
+  create(counter: ICounter, profile: string, container: string): ng.IPromise<any> {
+    counter["is-new-node"] = true;
+    return this.doCountersExist(container, profile).then((countersExists: boolean) => {
+      if (countersExists) {
+        return this.createHelper(counter, profile, container);
+      } else {
+        return this.createCountersDMRPath(container, profile).then((created: boolean) => {
+          if (created) {
+            return this.createHelper(counter, profile, container);
+          }
+        })
+      }
+    });
+  }
+
+  createHelper(counter: ICounter, profile: string, container: string): ng.IPromise<any> {
+    if (this.doCountersExist(container, profile)) {
+      if (counter.isStrong()) {
+        return this.save(counter, profile, container, ["is-new-node", "type", "concurrency"]);
+      } else {
+        return this.save(counter, profile, container, ["is-new-node", "type", "lower-bound", "upper-bound"]);
+      }
+    }
+  }
+
+  reset(container: ICacheContainer, counter: ICounter): ng.IPromise<any> {
+    let address: string[] = this.getCounterAddress(counter, container.name, container.profile);
+    return this.dmrService.executePost({
+      address: address,
+      operation: "counter-reset"
+    });
+  }
+
+  remove(container: ICacheContainer, counter: ICounter): ng.IPromise<any> {
+    let address: string[] = this.getCounterAddress(counter, container.name, container.profile);
+    return this.dmrService.executePost({
+      address: address,
+      operation: "remove"
+    });
+  }
+
+  save(counter: ICounter, profile: string, container: string, excludedAttributes: string []): ng.IPromise<any> {
+    let builder: CompositeOpBuilder = new CompositeOpBuilder();
+    let dmrAddress: string [] = this.getCounterAddress(counter, container, profile);
+    this.dmrService.traverseDMRTree(builder, counter, dmrAddress, excludedAttributes);
```

Comment:
I'm lost at this point. It's only a single add operation to create a counter. Why do we need to do all of this traversal? Can't we just keep it simple and do a simple add operation? e.g. https://github.com/infinispan/infinispan-management-console/blob/master/src/services/server/ServerService.ts#L30-L38

Reply:
Ok, I agree but I wanted to use this op that already creates all the necessary DMR operations for sava. Can we leave this for rework later?

Reply:
Ideally not as this is potentially creating more ops then required and it is something that will get forgot about. It should be a simple implementation, something like:

```typescript
createCounter(counter: ICounter, profile: string, container: string): ng.IPromise<void> {
    let request: IDmrRequest;
    if (counter instanceof StrongCounter) {
      let strongCounter: StrongCounter = <StrongCounter> counter;
      request = <IDmrRequest> {
        address: this.getCounterAddress(counter, container, profile),
        "initial-value": strongCounter.getInitialValue(),
        ...
      }
    } else {
      request = <IDmrRequest> {
        ...
      };
    }
    return this.dmrService.add(request);
  }
```